docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,269 @@
 | *By* | Imran M Yousuf |
 | *Date* | November 10, 2020 |
 
-## Problem Statement / Goals
+## Problem Statement / Goal
 
-In a service oriented or microservice architecture the necessity to reliable pass messages between systems is essential. We want a reliable message broker without increasing architectural complexity.
+In a service oriented or microservice architecture the necessity to reliably pass messages between systems is essential. We want a reliable message broker without increasing architectural complexity with high throughput and scalability.
+
+## Assumptions & Considerations
+
+The assumptions into this tech spec are -
+
+* All the consumers have an HTTP based interface available
+* The interface or data model of webhook payload is agreed upon between producer and consumer
+* The consumer will take care of out of order message being pushed
+
+The considerations are -
+
+* The broker will ensure at least once webhook event is delivered
+* Consumer should have access to all the messages it failed to receive due to error on its part
+* Consumer could access past messages for replaying purpose
+
+Key attributes deliberately out of consideration -
+
+* There will be no guarantee of message order
+* There is no consideration for Consumer errors caused by bugs
+
+## Use cases
+
+The obvious goal is to serve the purpose of Enterprise Service Bus over HTTP; it is best suited when ordering is not a necessity; for example, change streams with optimistic locking. This Broker is not intended to be used as a replacement for Kafka or AWS Kinesis when order sequence is absolutely necessary.
+
+### Yet another message broker
```

Comment:
I think that creating a webhook based broker is a novel idea because as your tech spec states, it removes the need to run a separate process to consumer messages from a broker. Your HTTP app becomes your worker and you don't need a second process so that's pretty cool. However we sacrifice a nice property of most message brokers which is that they will hold onto your messages and give them only when you ask for them. With the webhook based broker system it will constantly try to give you messages and I think this has some big downsides. My understanding of the delivery mechanism is that the webhook-broker makes an HTTP request to the consumer with the message, and when the consumer responds with HTTP 200 then the message is considered as delivered and then it can be removed from the queue. Will the webhook-broker still try to send messages to that consumer while it's busy? What if all the consumer processes are busy? I imagine that it will be really difficult to handle a large influx of messages with the webhook-broker approach vs the pull based approach of most message brokers. So I think that using an HTTP server as a worker seems like a simple idea but becomes complicated when you're faced with a lot of messages. The backlog of failed to deliver messages will time out. You could make the delivery and acknowledgement separate steps (deliver message, have consumer acknowledge via separate request), but you'll still have a backlog when trying to send new messages and then you'll need some async process in the broker to track how long the messages have been out for delivery to mark them as timed out.

Reply:
> Will the webhook-broker still try to send messages to that consumer while it's busy?

No the broker will not attempt to send the same message while it is being sent. In the life-cycle it is presented as - _Retry-In-Flight_ will block a message from being picked up and retrying being sent for the duration of `Consumer Timeout` setting. There are 2 scenario here -
1. The HTTP Client to Consumer triggers the timeout as it happens
1. The process crashes before receiving a response from the consumer.

In the first case after the timeout the status will be reinstated to _Retry-Delivery_ for the fail-safe mechanism to pick it up. Fail-safe mechanism will also run a second pass for _Retry-In-Flight_ that are older than Timeout + _Rational-Delay_ and try to process them as well; this will address the scenario 2.

> What if all the consumer processes are busy?

Valid concern, but not a deterrence to this approach because, it is fairly easy today to scale HTTP services, in fact it is far more easier. I always strongly felt that consumers should be able to aggressively scale rather than letting messages be stale in the queue. Solving the scalability of HTTP services is not a optional in any case; but now its simpler because you need to solve, test, maintain scalability of 1 type of service rather than multiple. So IMHO, it is overall simpler.

There are 2 other future techspecs after the initial version where consumer can say broker to stop pushing messages and/or set rate limit of broker pushing messages. When this is implemented, consumer can also choose to pull instead of receiving via push. That should also address this concern; but I would still prefer consumers solving the scalability problem.

> You could make the delivery and acknowledgement separate steps (deliver message, have consumer acknowledge via separate request)

There is 2 weakness of this approach - for both HTTP based or customer network protocol based. If you see how we had to tackle failure scenario for the first question, with separate ACK process now that scenario tracing will become more complicated. So I strongly feel message consumption signal should be within the same bounded network connection. Furthermore, I would love to avoid the consumer having to call the broker under normal circumstances.

Reply:
> No the broker will not attempt to send the same message while it is being sent.
What I meant here is if a consumer HTTP server is handling one message, when the broker gets a new message will it not immediately try to deliver the new message to the busy consumer HTTP server? This is the main problem I see with using an HTTP server as a consumer vs a process that pulls for messages when it's ready. Either way our workers are single threaded and if we want to scale to handle an increasing backlog we need to create more of them but there is no quick and easy mechanism for that (AWS autoscaling is slow). In addition the workload of this kind of process is different from a normal HTTP server which is normally just a few IO operations. Worker processes tend to have workloads that are a mix of CPU and lots of IO and generally take longer to complete than an HTTP request.

Reply:
Everything that you said is the bigger reason for choosing a HTTP driven approach :).

We will be sending another message to a consumer when it arrives without waiting for the earlier message to complete delivery. The assumption is that consumer should be ready to handle the message volume that it will receive; the capping initially will be the accumulated number of workers on the broker. We want the use of HTTP to be able to scale when queue grows larger; we need to solve the problem of how to scale fast.

Reply:
I agree we need to be clear that it is not for batch and stream processing but more for message passing.

Reply:
Add a section around how consumers should plan capacity on.